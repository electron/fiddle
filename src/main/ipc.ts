import { EventEmitter } from 'events';

import { ipcMain } from 'electron';

import {
  IpcEvents,
  WEBCONTENTS_READY_FOR_IPC_SIGNAL,
  ipcMainEvents,
} from '../ipc-events';
import { getOrCreateMainWindow } from './windows';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
class IpcMainManager extends EventEmitter {
  public readyWebContents = new WeakSet<Electron.WebContents>();
  private messageQueue = new WeakMap<
    Electron.WebContents,
    Array<[IpcEvents, Array<any> | undefined]>
  >();

  constructor() {
    super();

    ipcMainEvents.forEach((name) => {
      ipcMain.removeAllListeners(name);
      ipcMain.on(name, (...args: Array<any>) => this.emit(name, ...args));
    });

    ipcMain.on(
      WEBCONTENTS_READY_FOR_IPC_SIGNAL,
      (event: Electron.IpcMainEvent) => {
        this.readyWebContents.add(event.sender);

        const queue = this.messageQueue.get(event.sender);
        this.messageQueue.delete(event.sender);
        if (!queue) return;
        for (const item of queue) {
          this.send(item[0], item[1], event.sender);
        }
      },
    );
  }

  /**
   * Send an IPC message to an instance of Electron.WebContents.
   * If none is specified, we'll automatically go with the main window.
   *
   * @param {IpcEvents} channel
   * @param {Array<any>} [args]
   * @param {Electron.WebContents} [target]
   */
  public send(
    channel: IpcEvents,
    args?: Array<any>,
    target?: Electron.WebContents,
  ) {
    const _target = target || getOrCreateMainWindow().webContents;
    const _args = args || [];
    if (!this.readyWebContents.has(_target)) {
      const existing = this.messageQueue.get(_target) || [];
      this.messageQueue.set(_target, [...existing, [channel, args]]);
      return;
    }

    _target.isDestroyed() || _target.send(channel, ..._args);
  }

  public handle(
    channel: IpcEvents,
    listener: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any,
  ) {
    // there can be only one, so remove previous one first
    ipcMain.removeHandler(channel);
    ipcMain.handle(channel, listener);
  }

  public handleOnce(
    channel: IpcEvents,
    listener: (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any,
  ) {
    ipcMain.handleOnce(channel, listener);
  }
}

export const ipcMainManager = new IpcMainManager();
