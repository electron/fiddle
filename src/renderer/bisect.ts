import { RunnableVersion } from '../interfaces';

export class Bisector {
  public revList: Array<RunnableVersion>;
  public minRev: number;
  public maxRev: number;
  private pivot: number;

  constructor(revList: Array<RunnableVersion>) {
    this.getCurrentVersion = this.getCurrentVersion.bind(this);
    this.continue = this.continue.bind(this);
    this.calculatePivot = this.calculatePivot.bind(this);

    this.revList = revList;
    this.minRev = 0;
    this.maxRev = revList.length - 1;
    this.calculatePivot();
  }

  public getCurrentVersion() {
    return this.revList[this.pivot];
  }

  public continue(isGoodVersion: boolean) {
    let isBisectOver = false;
    if (this.maxRev - this.minRev <= 1) {
      isBisectOver = true;
    }

    if (isGoodVersion) {
      const upPivot = Math.floor((this.maxRev - this.pivot) / 2) + this.pivot;
      this.minRev = this.pivot;
      if (upPivot !== this.maxRev && upPivot !== this.pivot) {
        this.pivot = upPivot;
      } else {
        isBisectOver = true;
      }
    } else {
      const downPivot =
        Math.floor((this.pivot - this.minRev) / 2) + this.minRev;
      this.maxRev = this.pivot;
      if (downPivot !== this.minRev && downPivot !== this.pivot) {
        this.pivot = downPivot;
      } else {
        isBisectOver = true;
      }
    }

    if (isBisectOver) {
      return [this.revList[this.minRev], this.revList[this.maxRev]];
    } else {
      return this.revList[this.pivot];
    }
  }

  private calculatePivot() {
    this.pivot = Math.floor((this.maxRev - this.minRev) / 2);
  }
}
