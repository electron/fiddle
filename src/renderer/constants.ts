import * as path from 'path';

export const USER_DATA_PATH = window.ElectronFiddle.appPaths.userData;
export const ELECTRON_DOWNLOAD_PATH = path.join(USER_DATA_PATH, 'electron-bin');
export const ELECTRON_INSTALL_PATH = path.join(
  ELECTRON_DOWNLOAD_PATH,
  'current',
);
export const CONFIG_PATH = path.join(
  window.ElectronFiddle.appPaths.home,
  '.electron-fiddle',
);

export const ELECTRON_ORG = 'electron';
export const ELECTRON_REPO = 'electron';
