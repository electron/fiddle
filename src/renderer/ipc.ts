import { EventEmitter } from 'events';

import { ipcRenderer } from 'electron';

import { IpcEvents, ipcRendererEvents } from '../ipc-events';

/**
 * The main purpose of this class is to be the central
 * gathering place for IPC calls the main process sends
 * or listens to.
 *
 * @class IpcManager
 * @extends {EventEmitter}
 */
class IpcRendererManager extends EventEmitter {
  constructor() {
    super();

    ipcRendererEvents.forEach((name) => {
      ipcRenderer.removeAllListeners(name);
      ipcRenderer.on(name, (...args: Array<any>) => this.emit(name, ...args));
    });
  }

  /**
   * Central method to send an IPC message to the main process.
   * Currently very simple, but centralized here so that future
   * refactors will be a lot easier.
   *
   * @param {IpcEvents} channel
   * @param {...Array<any>} args
   * @memberof IpcRendererManager
   */
  public send(channel: IpcEvents, ...args: Array<any>) {
    ipcRenderer.send(channel, ...args);
  }

  public invoke(channel: IpcEvents, ...args: Array<any>) {
    return ipcRenderer.invoke(channel, ...args);
  }
}

export const ipcRendererManager = new IpcRendererManager();
