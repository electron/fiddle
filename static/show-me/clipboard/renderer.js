window.addEventListener('DOMContentLoaded', () => {
  const copyButton = document.querySelector('#copy')
  const pasteButton = document.querySelector('#paste')
  const textarea = document.querySelector('textarea')
  copyButton.onclick = () => {
    window.clipboard.writeText('Hello from Electron!')
  }
  pasteButton.onclick = async () => {
    textarea.value = await window.clipboard.readText()
  }
})
