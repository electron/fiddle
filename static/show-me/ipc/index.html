<!DOCTYPE html>
<html lang="en">
<head>
  <title>IPC Example</title>
</head>
<body>
  <h1>IPC Example</h1>
  <p>Open the developer tools to see the IPC in action.</p>
</body>
</html>
