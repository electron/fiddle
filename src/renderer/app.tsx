import * as path from 'path';

import { ipcRenderer } from 'electron';
import { autorun, reaction, when } from 'mobx';

import { EditorValues, PACKAGE_NAME, SetFiddleOptions } from '../interfaces';
import { IpcEvents, WEBCONTENTS_READY_FOR_IPC_SIGNAL } from '../ipc-events';
import { PackageJsonOptions, getPackageJson } from '../utils/get-package';
import { USER_DATA_PATH } from './constants';
import { ElectronTypes } from './electron-types';
import { FileManager } from './file-manager';
import { ipcRendererManager } from './ipc';
import { RemoteLoader } from './remote-loader';
import { Runner } from './runner';
import { AppState } from './state';
import { TaskRunner } from './task-runner';
import { activateTheme, getTheme } from './themes';
import { defaultDark, defaultLight } from './themes-defaults';
import { getElectronVersions } from './versions';

// Importing styles files
import '../less/root.less';

/**
 * The top-level class controlling the whole app. This is *not* a React component,
 * but it does eventually render all components.
 *
 * @class App
 */
export class App {
  public state = new AppState(getElectronVersions());
  public fileManager = new FileManager(this.state);
  public remoteLoader = new RemoteLoader(this.state);
  public runner = new Runner(this.state);
  public readonly taskRunner: TaskRunner;
  public readonly electronTypes: ElectronTypes;

  constructor() {
    this.getEditorValues = this.getEditorValues.bind(this);

    this.taskRunner = new TaskRunner(this);

    this.electronTypes = new ElectronTypes(
      window.ElectronFiddle.monaco,
      path.join(USER_DATA_PATH, 'electron-typedef'),
      path.join(USER_DATA_PATH, 'nodejs-typedef'),
    );
  }

  private confirmReplaceUnsaved(): Promise<boolean> {
    return this.state.showConfirmDialog({
      label: `Opening this Fiddle will replace your unsaved changes. Do you want to proceed?`,
      ok: 'Open',
    });
  }

  private confirmExitUnsaved(): Promise<boolean> {
    return this.state.showConfirmDialog({
      label: 'The current Fiddle is unsaved. Do you want to exit anyway?',
      ok: 'Exit',
    });
  }

  public async replaceFiddle(
    editorValues: EditorValues,
    { filePath, gistId, templateName }: Partial<SetFiddleOptions>,
  ) {
    const { state } = this;
    const { editorMosaic } = state;

    if (editorMosaic.isEdited && !(await this.confirmReplaceUnsaved())) {
      return false;
    }

    this.state.editorMosaic.set(editorValues);

    this.state.gistId = gistId || '';
    this.state.localPath = filePath;
    this.state.templateName = templateName;

    // update menu when a new Fiddle is loaded
    ipcRenderer.send(IpcEvents.SET_SHOW_ME_TEMPLATE, templateName);

    return true;
  }

  /**
   * Retrieves the contents of all editor panes.
   *
   * @returns {EditorValues}
   */
  public async getEditorValues(
    options?: PackageJsonOptions,
  ): Promise<EditorValues> {
    const values = this.state.editorMosaic.values();

    if (options) {
      values[PACKAGE_NAME] = await getPackageJson(this.state, options);
    }

    return values;
  }

  /**
   * Initial setup call, loading Monaco and kicking off the React
   * render process.
   */
  public async setup(): Promise<void | Element | React.Component> {
    this.loadTheme(this.state.theme || '');

    const React = await import('react');
    const { render } = await import('react-dom');
    const { Dialogs } = await import('./components/dialogs');
    const { OutputEditorsWrapper } = await import(
      './components/output-editors-wrapper'
    );
    const { Header } = await import('./components/header');

    // The AppState constructor started loading a fiddle.
    // Wait for it here so the UI doesn't start life in `nonIdealState`.
    await when(() => this.state.editorMosaic.files.size !== 0);

    const className = `${process.platform} container`;
    const app = (
      <div className={className}>
        <Dialogs appState={this.state} />
        <Header appState={this.state} />
        <OutputEditorsWrapper appState={this.state} />
      </div>
    );

    const rendered = render(app, document.getElementById('app'));

    this.setupResizeListener();
    this.setupOfflineListener();
    this.setupThemeListeners();
    this.setupTitleListeners();
    this.setupUnloadListeners();
    this.setupTypeListeners();

    ipcRenderer.send(WEBCONTENTS_READY_FOR_IPC_SIGNAL);

    ipcRenderer.on(IpcEvents.SET_SHOW_ME_TEMPLATE, () => {
      ipcRenderer.send(IpcEvents.SET_SHOW_ME_TEMPLATE, this.state.templateName);
    });

    return rendered;
  }

  private setupTypeListeners() {
    const updateTypes = () =>
      this.electronTypes.setVersion(this.state.currentElectronVersion);
    reaction(
      () => this.state.version,
      () => updateTypes(),
    );
    updateTypes();
  }

  public async setupThemeListeners() {
    const setSystemTheme = (prefersDark: boolean) => {
      if (prefersDark) {
        this.state.setTheme(defaultDark.file);
      } else {
        this.state.setTheme(defaultLight.file);
      }
    };

    // match theme to system when box is ticked
    reaction(
      () => this.state.isUsingSystemTheme,
      () => {
        if (this.state.isUsingSystemTheme && !!window.matchMedia) {
          const { matches } = window.matchMedia('(prefers-color-scheme: dark)');
          setSystemTheme(matches);
        }
      },
    );

    // change theme when system theme changes
    if (!!window.matchMedia) {
      window
        .matchMedia('(prefers-color-scheme: dark)')
        .addEventListener('change', ({ matches }) => {
          if (this.state.isUsingSystemTheme) {
            setSystemTheme(matches);
          }
        });
    }
  }

  /**
   * Opens a fiddle from the specified location.
   *
   * @param {SetFiddleOptions} fiddle The fiddle to open
   */
  public async openFiddle(fiddle: SetFiddleOptions) {
    const { filePath, gistId } = fiddle;
    if (filePath) {
      await this.fileManager.openFiddle(filePath);
    } else if (gistId) {
      await this.remoteLoader.fetchGistAndLoad(gistId);
    }
  }

  /**
   * Loads theme CSS into the HTML document.
   *
   * @param {string} name
   * @returns {Promise<void>}
   */
  public async loadTheme(name: string): Promise<void> {
    const tag: HTMLStyleElement | null = document.querySelector(
      'style#fiddle-theme',
    );
    const theme = await getTheme(name);
    activateTheme(theme);

    if (tag && theme.css) {
      tag.innerHTML = theme.css;
    }

    if (theme.isDark || theme.name.includes('dark')) {
      document.body.classList.add('bp3-dark');
    } else {
      document.body.classList.remove('bp3-dark');
    }
  }

  public setupOfflineListener(): void {
    window.addEventListener('online', async () => {
      this.state.isOnline = true;
      this.state.setVersion(this.state.version);
    });
    window.addEventListener('offline', () => {
      this.state.isOnline = false;
    });
  }

  /**
   * We need to possibly recalculate the layout whenever the window
   * is resized. This method sets up the listener.
   */
  public setupResizeListener(): void {
    window.addEventListener('resize', this.state.editorMosaic.layout);
  }

  /**
   * Have document.title track state.title
   */
  public setupTitleListeners() {
    // the observables used for the title usually change in a batch,
    // so when setting document title, wait a tick to avoid flicker.
    let titleIdle: any;
    reaction(
      () => this.state.title,
      (title) => {
        clearTimeout(titleIdle);
        titleIdle = setTimeout(() => {
          document.title = title;
          titleIdle = null;
        });
      },
    );
  }

  public setupUnloadListeners() {
    autorun(async () => {
      const { state } = this;
      const { editorMosaic } = state;

      if (!editorMosaic.isEdited) {
        window.onbeforeunload = null;
        return;
      }

      window.onbeforeunload = async () => {
        if (await this.confirmExitUnsaved()) {
          // isQuitting checks if we're trying to quit the app
          // or just close the window
          if (state.isQuitting) {
            ipcRendererManager.send(IpcEvents.CONFIRM_QUIT);
          }
          window.onbeforeunload = null;
          window.close();
        }

        // return value doesn't matter, we just want to cancel the event
        return false;
      };
    });
  }
}
