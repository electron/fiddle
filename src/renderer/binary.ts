import * as fsType from 'fs-extra';
import * as path from 'path';
import * as extractZipType from 'extract-zip';

import { VersionState } from '../interfaces';
import { fancyImport } from '../utils/import';
import { normalizeVersion } from '../utils/normalize-version';
import { USER_DATA_PATH } from './constants';
import { removeTypeDefsForVersion } from './fetch-types';
import { AppState } from './state';

const pendingDownloads: Record<string, Promise<void>> = {};

async function downloadBinary(appState: AppState, version: string): Promise<void> {
  console.log(`Binary: Electron ${version} not present, downloading`);
  appState.versions[version].state = VersionState.downloading;

  const zipPath = await download(appState, version);
  const extractPath = getDownloadPath(version);
  console.log(
    `Binary: Electron ${version} downloaded, now unpacking to ${extractPath}`,
  );

  try {
    appState.versions[version].state = VersionState.unzipping;

    const fs = await fancyImport<typeof fsType>('fs-extra');
    await fs.mkdirp(getDownloadPath(version));
    await fs.emptyDir(extractPath);
    const electronFiles = await unzip(zipPath, extractPath);
    console.log(`Binary: Unzipped ${version}`, electronFiles);
    appState.versions[version].state = VersionState.ready;
  } catch (error) {
    console.warn(`Binary: Failure while unzipping ${version}`, error);
    appState.versions[version].state = VersionState.unknown;
  } finally {
    // This task is done, so remove it from the pending tasks list
    delete pendingDownloads[version];
  }
}

/**
 * General setup, called with a version. Is called during construction
 * to ensure that we always have or download at least one version.
 *
 * @param {string} iVersion
 * @returns {Promise<void>}
 */
export async function setupBinary(
  appState: AppState,
  iVersion: string,
): Promise<void> {
  const version = normalizeVersion(iVersion);

  // If we already have it, then we're done
  if (await getIsDownloaded(version)) {
    console.log(`Binary: Electron ${version} already downloaded.`);
    appState.versions[version].state = VersionState.ready;
    return;
  }

  // Return a promise that resolves when the download completes
  let pending = pendingDownloads[version];
  if (pending) {
    console.log(`Binary: Electron ${version} already downloading.`);
  } else {
    pending = pendingDownloads[version] = downloadBinary(appState, version);
  }
  return pending;
}

/**
 * Remove a version from disk. Does not update state. We'll try up to
 * four times before giving up if an error occurs.
 *
 * @param {string} iVersion
 * @returns {Promise<void>}
 */
export async function removeBinary(iVersion: string) {
  const version = normalizeVersion(iVersion);
  const fs = await fancyImport<typeof fsType>('fs-extra');
  let isDeleted = false;

  // utility to re-run removal functions upon failure
  // due to windows filesystem lockfile jank
  const rerunner = async (func: () => Promise<void>, counter = 1) => {
    try {
      await func();
    } catch (error) {
      console.warn(
        `Binary Manager: failed to run ${func.name} for ${version}, but failed`,
        error,
      );
      if (counter < 4) {
        console.log(`Binary Manager: Trying again to run ${func.name}`);
        await rerunner(func, counter + 1);
      }
    }
  };

  const binaryCleaner = async () => {
    if (await getIsDownloaded(version)) {
      // This is necessary since we're messing with .asar files inside
      // the Electron binaries. Electron, powering Fiddle, will try to
      // "correct" our calls, but we don't want that right here.
      process.noAsar = true;
      await fs.remove(getDownloadPath(version));
      process.noAsar = false;

      isDeleted = true;
    }
  };

  const typeDefsCleaner = async () => {
    await removeTypeDefsForVersion(version);
  };

  await rerunner(binaryCleaner);

  if (isDeleted) {
    await rerunner(typeDefsCleaner);
  }
}

/* Did we already download a given version?
 *
 * @param {string} version
 * @param {string} dir
 * @returns {boolean}
 */
export async function getIsDownloaded(
  version: string,
  dir?: string,
): Promise<boolean> {
  const expectedPath = getElectronBinaryPath(version, dir);
  const fs = await fancyImport<typeof fsType>('fs-extra');

  return fs.existsSync(expectedPath);
}

/**
 * Gets the expected path for the binary of a given Electron version
 *
 * @param {string} version
 * @param {string} dir
 * @returns {string}
 */
export function getElectronBinaryPath(
  version: string,
  dir: string = getDownloadPath(version),
): string {
  switch (process.platform) {
    case 'darwin':
      return path.join(dir, 'Electron.app/Contents/MacOS/Electron');
    case 'freebsd':
    case 'linux':
      return path.join(dir, 'electron');
    case 'win32':
      return path.join(dir, 'electron.exe');
    default:
      throw new Error(
        `Electron builds are not available for ${process.platform}`,
      );
  }
}

export function getDownloadingVersions(appState: AppState) {
  return Object.entries(appState.versions)
    .filter(([_, { state }]) => state === 'downloading')
    .map(([version, _]) => version);
}

/**
 * Returns an array of all versions downloaded to disk
 *
 * @returns {Promise<Array<string>>}
 */
export async function getDownloadedVersions(): Promise<Array<string>> {
  const fs = await fancyImport<typeof fsType>('fs-extra');
  const downloadPath = path.join(USER_DATA_PATH, 'electron-bin');
  console.log(`Binary: Checking for downloaded versions`);

  try {
    const directories = await fs.readdir(downloadPath);
    const knownVersions: Array<string> = [];

    for (const directory of directories) {
      if (await getIsDownloaded(directory)) {
        knownVersions.push(directory);
      }
    }

    return knownVersions;
  } catch (error) {
    console.warn(`Could not read known Electron versions`);
    return [];
  }
}

/**
 * Download an Electron version.
 *
 * @param {AppState} appState
 * @param {string} version
 * @returns {Promise<string>}
 */
async function download(appState: AppState, version: string): Promise<string> {
  const { download: electronDownload } = await import('@electron/get');
  const getProgressCallback = (progress: Progress) => {
    const roundedProgress = Math.round(progress.percent * 100) / 100;

    if (roundedProgress !== appState.versions[version].downloadProgress) {
      console.debug(
        `Binary: Version ${version} download progress: ${progress.percent}`,
      );
      appState.versions[version].downloadProgress = roundedProgress;
    }
  };

  if (!appState.versions[version]) {
    throw new Error(
      `Version ${version} does not exist in state, cannot download`,
    );
  }

  const zipFilePath = await electronDownload(version, {
    downloadOptions: {
      quiet: true,
      getProgressCallback,
    },
  });

  return zipFilePath;
}

/**
 * Gets the expected path for a given Electron version
 *
 * @param {string} version
 * @returns {string}
 */
function getDownloadPath(version: string): string {
  return path.join(USER_DATA_PATH, 'electron-bin', version);
}

/**
 * Unzips an electron package so that we can actually use it.
 *
 * @param {string} zipPath
 * @param {string} dir
 * @returns {Promise<void>}
 */
async function unzip(zipPath: string, dir: string): Promise<void> {
  const extract: typeof extractZipType = (await fancyImport<any>('extract-zip'))
    .default;

  process.noAsar = true;

  try {
    await extract(zipPath, { dir });
    console.log(`Binary: Unpacked!`);
  } finally {
    process.noAsar = false;
  }
}

interface Progress {
  percent: number;
  transferred: number;
  total: number;
}
