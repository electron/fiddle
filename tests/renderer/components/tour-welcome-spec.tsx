import * as React from 'react';

import { shallow } from 'enzyme';

import {
  WelcomeTour,
  getWelcomeTour,
} from '../../../src/renderer/components/tour-welcome';
import { ipcRendererManager } from '../../../src/renderer/ipc';
import { StateMock } from '../../mocks/mocks';

describe('Header component', () => {
  let store: StateMock;

  beforeEach(() => {
    ({ state: store } = (window as any).ElectronFiddle.app);
    store.isTourShowing = true;

    ipcRendererManager.removeAllListeners();
  });

  it('renders', () => {
    const wrapper = shallow(<WelcomeTour appState={store as any} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders null if the tour is not showing', () => {
    store.isTourShowing = false;

    const wrapper = shallow(<WelcomeTour appState={store as any} />);
    expect(wrapper.html()).toBe(null);
  });

  it('renders the tour once started', () => {
    const wrapper = shallow(<WelcomeTour appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    instance.startTour();

    expect(wrapper.state('isTourStarted')).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('stops the tour on stopTour()', () => {
    const wrapper = shallow(<WelcomeTour appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    instance.stopTour();

    expect(wrapper.state('isTourStarted')).toBe(false);
    expect(store.disableTour).toHaveBeenCalled();
  });

  describe('getWelcomeTour()', () => {
    it('offers custom buttons for the Electron step', () => {
      const tourSteps = [...getWelcomeTour()];
      const electronStep = tourSteps.find(
        ({ name }) => name === 'first-time-electron',
      );
      const mockParam = { stop: jest.fn(), advance: jest.fn() };
      const buttons = electronStep!.getButtons!(mockParam);

      shallow(buttons[0]).simulate('click');
      expect(mockParam.stop).toHaveBeenCalled();

      shallow(buttons[1]).simulate('click');
      expect(mockParam.advance).toHaveBeenCalled();
    });
  });
});
