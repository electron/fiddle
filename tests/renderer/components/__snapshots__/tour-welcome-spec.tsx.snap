// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders 1`] = `
<Blueprint3.Dialog
  canOutsideClickClose={true}
  isOpen={true}
  key="welcome-tour-dialog"
>
  <div
    className="bp3-dialog-header"
  >
    <h4
      className="bp3-heading"
    >
      üôã‚Äç Hey There!
    </h4>
  </div>
  <div
    className="bp3-dialog-body"
  >
    <p>
      Welcome to Electron Fiddle! If you're new to the app, we'd like to give you a brief tour of its features.
    </p>
    <p>
      We won't show this dialog again, but you can always find the tour in the Help menu.
    </p>
  </div>
  <div
    className="bp3-dialog-footer"
  >
    <div
      className="bp3-dialog-footer-actions"
    >
      <Blueprint3.Button
        icon="cross"
        key="cancel"
        onClick={[Function]}
        text="I'll figure it out"
      />
      <Blueprint3.Button
        icon="presentation"
        key="ok"
        onClick={[Function]}
        text="Show me around"
      />
    </div>
  </div>
</Blueprint3.Dialog>
`;

exports[`Header component renders the tour once started 1`] = `
<Tour
  onStop={[Function]}
  tour={
    Set {
      Object {
        "content": <React.Fragment>
          <p>
            Electron Fiddle allows you to build little experiments and mini-apps with Electron. Each Fiddle has at least three of these files: A main script, a renderer script, a preload script, and an HTML file.
          </p>
          <p>
            If you 
            <code>
              require()
            </code>
             a module, Fiddle will install it automatically. It will also automatically provide you with autocomplete information for the 
            <code>
              electron
            </code>
             module.
          </p>
        </React.Fragment>,
        "name": "fiddle-editors",
        "selector": "div.mosaic-root",
        "title": "üìù Fiddle Editors",
      },
      Object {
        "content": <React.Fragment>
          <p>
            Electron Fiddle knows about all released Electron versions, downloading your versions automatically in the background.
          </p>
          <p>
            Open the preferences to see all available versions and delete those previously downloaded.
          </p>
        </React.Fragment>,
        "name": "select-versions",
        "selector": "button.version-chooser",
        "title": "üìá Choose an Electron Version",
      },
      Object {
        "content": <p>
          Hit this button to give your Fiddle a try and start it.
        </p>,
        "name": "button-run",
        "selector": ".button-run",
        "title": "üöÄ Run Your Fiddle",
      },
      Object {
        "content": <React.Fragment>
          <p>
            Like what you've built? You can save your Fiddle as a public GitHub Gist, allowing other users to load it by pasting the URL into the address bar. If they don't have Electron Fiddle, they can see and download your code directly from GitHub.
          </p>
          <p>
            You can also package your Fiddle as a standalone binary or as an installer from the "Tasks" menu.
          </p>
        </React.Fragment>,
        "name": "button-publish",
        "selector": ".button-publish",
        "title": "üó∫ Share Your Fiddle",
      },
      Object {
        "content": <p>
          We've finished our tour of Electron Fiddle, but if this is your first time using Electron, we could introduce you to its basics. Interested?
        </p>,
        "getButtons": [Function],
        "name": "first-time-electron",
        "selector": "div.mosaic-root",
        "title": "üëã Getting Started With Electron?",
      },
      Object {
        "content": <React.Fragment>
          <p>
            Every Electron app starts with a main script, very similar to how a Node.js application is started. The main script runs in the "main process". To display a user interface, the main process creates renderer processes ‚Äì¬†usually in the form of windows, which Electron calls ¬†
            <code>
              BrowserWindow
            </code>
            .
          </p>
          <p>
            To get started, pretend that the main process is just like a Node.js process. All APIs and features found in Electron are accessible through the 
            <code>
              electron
            </code>
             module, which can be required like any other Node.js module.
          </p>
          <p>
            The default fiddle creates a new 
            <code>
              BrowserWindow
            </code>
             and loads an HTML file.
          </p>
        </React.Fragment>,
        "name": "main-editor",
        "selector": "div.mosaic-window.main",
        "title": "üìù Main Script",
      },
      Object {
        "content": <p>
          In the default fiddle, this HTML file is loaded in the ¬†
          <code>
            BrowserWindow
          </code>
          . Any HTML, CSS, or JavaScript that works in a browser will work here, too. In addition, Electron allows you to execute Node.js code. Take a close look at the ¬†
          <code>
            &lt;script /&gt;
          </code>
           tag and notice how we can call
           
          <code>
            require()
          </code>
           like we would in Node.js.
        </p>,
        "name": "html-editor",
        "selector": "div.mosaic-window.html",
        "title": "üìù HTML",
      },
      Object {
        "content": <React.Fragment>
          <p>
            This is the script we just required from the HTML file. In here, you can do anything that works in Node.js 
            <i>
              and
            </i>
             anything that works in a browser.
          </p>
          <p>
            By the way: If you want to use an 
            <code>
              npm
            </code>
             module here, just ¬†
            <code>
              require
            </code>
             it. Electron Fiddle will automatically detect that you requested a module and install it as soon as you run your fiddle.
          </p>
        </React.Fragment>,
        "name": "renderer-editor",
        "selector": "div.mosaic-window.renderer",
        "title": "üìù  Renderer Script",
      },
    }
  }
/>
`;
